feat: add transform option

hansottowirtz opened the pull request:
## Motivation

My use case is as follows:

```tsx
function Wrapper({ node, children }) {
  const [show, setShow] = useState(false)
  
  return <>
    {show && children}
    <button onClick={() => setShow(s => !s)}>Toggle</button>
  </>
}

domToReact(nodes, {
  transform(node, reactNode) {
    return <Wrapper node={node}>{reactNode}</Wrapper>
  }
})
```

I want to wrap every element in a Wrapper. I don't think this is possible with `replace` as it "replaces" the elements, not wraps them, and I can't easily "pass through" the subtree.

If you think this PR is worth considering I will follow up with tests and extra documentation.

## Checklist:

- [x] [Conventional Commits](https://www.conventionalcommits.org/)
- [x] Tests
- [x] Documentation
- [x] Types

remarkablemark reviewed (changes requested):
@hansottowirtz Thanks for opening this PR! Do you mind amending your commit message so it passes commitlint?

```sh
git commit --amend -m "feat: add transform option"
```

Also, could you add tests and documentation?

remarkablemark reviewed (commented):
Looks good! Could you also document this option in `README.md`?

hansottowirtz commented:
Sorry for the many pushes, if I need to write more tests or documentation let me know :) I changed the order of arguments as well (reactNode, node, index) and added the index.

remarkablemark commented on lib/dom-to-react.js:
```diff
@@ -26,6 +27,11 @@ function domToReact(nodes, options) {
   var hasReplace = typeof options.replace === 'function';
+  var transform =
+    options.transform ||
+    function (reactNode) {
+      return reactNode;
+    };
```
Minor: can we create a `noop` function in `utilities` as a performance optimization?

hansottowirtz replied:
Sure, however a noop would be `function noop(arg) {}` in my opinion, is `function returnFirstArg(arg) { return arg }` also fine?

remarkablemark replied:
Yes perfect

remarkablemark commented on README.md:
````diff
@@ -303,6 +304,21 @@ HTML output:
+
+```jsx
+parse('<br>', {
+  transform: (reactNode, node, index) => {
````
Nit: can we rename `node` to `domNode`?

```suggestion
  transform: (reactNode, domNode, index) => {
```

remarkablemark reviewed (dismissed):
Thanks for addressing the comments. I left a few more minor/nitpick comments.

remarkablemark reviewed (approved):
LGTM

remarkablemark commented:
Release:

- https://github.com/remarkablemark/html-react-parser/pull/988#issuecomment-1644473963
